tmail-backend/integration-tests/rate-limiter/rate-limiter-integration-tests-common/src/main/scala/com/linagora/tmail/integration/RateLimitingPlanIntegrationContract.scala

```diff
@@ -0,0 +1,179 @@
+
+    messageSender(server).authenticate(SENDER1.asString(), PASSWORD)
+      .sendMessageWithHeaders(SENDER1.asString, RECIPIENT1.asString, "subject: test123\r\rcontent 123\r.\r")
+
+    val imapClient: TestIMAPClient = testImapClient(server)
+    imapClient.login(RECIPIENT1.asString(), PASSWORD)
+      .select(INBOX)
+      .awaitMessage(awaitAtMostOneMinute)
+
+    assertThat(imapClient.readFirstMessage).contains("content 123")
+  }
+
+  @Test
+  def senderShouldNotSentEmailWhenRateLimitPlanIsExceeded(server: GuiceJamesServer): Unit = {
+    attachPlanToUser(createNewPlan(), SENDER1)
+
+    // accept
+    messageSender(server).authenticate(SENDER1.asString(), PASSWORD)
+      .sendMessageWithHeaders(SENDER1.asString, RECIPIENT1.asString, "subject: test1\r\rcontent 1\r.\r")
+
+    testImapClient(server).login(RECIPIENT1.asString(), PASSWORD)
+      .select(INBOX)
+      .awaitMessage(awaitAtMostOneMinute)
+
+    // exceed
+    messageSender(server).authenticate(SENDER1.asString(), PASSWORD)
+      .sendMessageWithHeaders(SENDER1.asString, RECIPIENT2.asString, "subject: test2\r\rcontent 2\r.\r")
+
+    awaitAtMostOneMinute.until(() => server.getProbe(classOf[MailRepositoryProbeImpl])
+      .getRepositoryMailCount(ERROR_REPOSITORY) == 1)
+
+    assertThat(testImapClient(server)
+      .login(RECIPIENT2, PASSWORD)
+      .getMessageCount(INBOX))
+      .isEqualTo(0)
+  }
+
+  @Test
+  def recipientShouldReceivedEmailWhenRateLimitPlanIsAcceptable(server: GuiceJamesServer): Unit = {
+    attachPlanToUser(createNewPlan(), RECIPIENT1)
+
+    messageSender(server).authenticate(SENDER1.asString(), PASSWORD)
+      .sendMessageWithHeaders(SENDER1.asString, RECIPIENT1.asString, "subject: test123\r\rcontent 123\r.\r")
+
+    val imapClient: TestIMAPClient = testImapClient(server)
+    imapClient.login(RECIPIENT1.asString(), PASSWORD)
+      .select(INBOX)
+      .awaitMessage(awaitAtMostOneMinute)
+
+    assertThat(imapClient.readFirstMessage).contains("content 123")
+  }
+
+  @Test
+  def recipientShouldNotReceivedEmailWhenRateLimitPlanIsExceeded(server: GuiceJamesServer): Unit = {
+    attachPlanToUser(createNewPlan(), RECIPIENT1)
+    // accept
+    messageSender(server).authenticate(SENDER1.asString(), PASSWORD)
+      .sendMessageWithHeaders(SENDER1.asString, RECIPIENT1.asString, "subject: test1\r\rcontent 1\r.\r")
+
+    testImapClient(server).login(RECIPIENT1.asString(), PASSWORD)
+      .select(INBOX)
+      .awaitMessage(awaitAtMostOneMinute)
+
+    // exceed
+    messageSender(server).authenticate(SENDER1.asString(), PASSWORD)
+      .sendMessageWithHeaders(SENDER1.asString, RECIPIENT1.asString, "subject: test2\r\rcontent 2\r.\r")
+
+    awaitAtMostOneMinute.until(() => server.getProbe(classOf[MailRepositoryProbeImpl])
+      .getRepositoryMailCount(ERROR_REPOSITORY) == 1)
+
+    assertThat(testImapClient(server)
+      .login(RECIPIENT1, PASSWORD)
+      .getMessageCount(INBOX))
+      .isEqualTo(1)
+  }
+
```

Comment:
Should we have a test for applying different rate limit plans to different users?